File: app/src/main/java/com/chuongvd/app/signal/data/source/repository/SignalRepository.java
package com.chuongvd.app.signal.data.source.repository;

import android.arch.lifecycle.LiveData;
import com.chuongvd.app.signal.BasicApp;
import com.chuongvd.app.signal.data.database.AppDatabase;
import com.chuongvd.app.signal.data.entity.SignalEntity;
import com.chuongvd.app.signal.data.source.local.datasource.ISignalLocalDataSource;
import com.chuongvd.app.signal.data.source.local.datasource.SignalLocalDataSource;
import com.chuongvd.app.signal.data.source.remote.datasource.ISignalRemoteDataSource;
import com.chuongvd.app.signal.data.source.remote.datasource.SignalRemoteDataSource;
import io.reactivex.Observable;
import java.util.List;

public class SignalRepository implements ISignalLocalDataSource, ISignalRemoteDataSource {

    private static SignalRepository sInstance;
    private ISignalRemoteDataSource mRemoteDataSource;
    private ISignalLocalDataSource mLocalDataSource;

    private SignalRepository(SignalRemoteDataSource remoteDataSource,
            SignalLocalDataSource localDataSource) {
        mRemoteDataSource = remoteDataSource;
        mLocalDataSource = localDataSource;
    }

    public static SignalRepository getInstance() {
        if (sInstance == null) {
            synchronized (SignalRepository.class) {
                if (sInstance == null) {
                    sInstance = new SignalRepository(SignalRemoteDataSource.getInstance(),
                            SignalLocalDataSource.getInstance(
                                    AppDatabase.getInstance(BasicApp.self())));
                }
            }
        }
        return sInstance;
    }

    @Override
    public void saveSignal(SignalEntity signalEntity) {
        mLocalDataSource.saveSignal(signalEntity);
    }

    @Override
    public void saveListSignal(List<SignalEntity> list) {
        mLocalDataSource.saveListSignal(list);
    }

    @Override
    public LiveData<List<SignalEntity>> getLocalSignals() {
        return mLocalDataSource.getLocalSignals();
    }

    @Override
    public Observable<List<SignalEntity>> getRemoteListSignal() {
        return mRemoteDataSource.getRemoteListSignal();
    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/repository/SignalRepository.kt
package com.chuongvd.app.signal.data.source.repository

import android.arch.lifecycle.LiveData
import com.chuongvd.app.signal.BasicApp
import com.chuongvd.app.signal.data.database.AppDatabase
import com.chuongvd.app.signal.data.entity.SignalEntity
import com.chuongvd.app.signal.data.source.local.datasource.ISignalLocalDataSource
import com.chuongvd.app.signal.data.source.local.datasource.SignalLocalDataSource
import com.chuongvd.app.signal.data.source.remote.datasource.ISignalRemoteDataSource
import com.chuongvd.app.signal.data.source.remote.datasource.SignalRemoteDataSource
import io.reactivex.Observable

class SignalRepository private constructor(remoteDataSource: SignalRemoteDataSource,
    localDataSource: SignalLocalDataSource) : ISignalLocalDataSource, ISignalRemoteDataSource {
  private val mRemoteDataSource: ISignalRemoteDataSource
  private val mLocalDataSource: ISignalLocalDataSource

  override val localSignals: LiveData<List<SignalEntity>>
    get() = mLocalDataSource.localSignals

  override val remoteListSignal: Observable<List<SignalEntity>>
    get() = mRemoteDataSource.remoteListSignal

  init {
    mRemoteDataSource = remoteDataSource
    mLocalDataSource = localDataSource
  }

  override fun saveSignal(signalEntity: SignalEntity) {
    mLocalDataSource.saveSignal(signalEntity)
  }

  override fun saveListSignal(list: List<SignalEntity>) {
    mLocalDataSource.saveListSignal(list)
  }

  companion object {

    private var sInstance: SignalRepository? = null

    val instance: SignalRepository
      get() {
        if (sInstance == null) {
          synchronized(SignalRepository::class.java) {
            if (sInstance == null) {
              sInstance = SignalRepository(SignalRemoteDataSource.instance,
                  SignalLocalDataSource.getInstance(
                      AppDatabase.getInstance(BasicApp.self)))
            }
          }
        }
        return this.sInstance!!
      }
  }
}
